draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1439,159 +1538,248 @@ Steps:
+- len\_in\_octets, the number of octets to be generated.
+
+This document defines two variants of expand\_message:
+
+- expand\_message\_md ({{hashtofield-expand-md}}) is appropriate for use
+with a wide range of hash functions, including SHA-2 {{FIPS180-4}}, SHA-3
+{{FIPS202}}, BLAKE2 {{?RFC7693}}, and others.
+
+- expand\_message\_xof ({{hashtofield-expand-xof}}) is appropriate for use
+with extensible-output functions (XOFs) including functions in the SHAKE
+{{FIPS202}} or BLAKE2X {{BLAKE2X}} families.
+
+These two variants should suffice for the vast majority of use cases, but other
+variants are possible; {{hashtofield-expand-other}} discusses requirements.
+
+### expand\_message\_md {#hashtofield-expand-md}
+
+The expand\_message\_md function produces a pseudorandom octet string using
+a cryptographic hash function H that outputs b bits.
+For security, H must meet the following requirements:
+
+- The number of bits output by H MUST be b >= 2 * k, for k the target
+security level in bits. This ensures k-bit collision resistance.
+
+- H MAY be a Merkle-Damgaard hash function like SHA-2.
+In this case, security holds when the underlying compression function is
+modeled as a random oracle {{CDMP05}}.
+(See {{security-considerations-expand-md}} for discussion.)
+
+- H MAY be a sponge-based hash function like SHA-3 or BLAKE2.
+In this case, security holds when the inner function is modeled as a
+random transformation or as a random permutation {{BDPV08}}.
+
+- Otherwise, H MUST be a hash function that has been proved indifferentiable
+from a random oracle {{MRH04}} under a reasonable cryptographic assumption.
+
+SHA-2 {{FIPS180-4}} and SHA-3 {{FIPS202}} are typical and RECOMMENDED choices.
+As an example, for 128-bit security, b >= 256 bits and either SHA-256 or
+SHA3-256 would be an appropriate choice.
+
+The following procedure implements expand\_message\_md.
+
+~~~
+expand_message_md(msg, DST, len_in_octets)
+
+Parameters:
+- H, a hash function (see requirements above).
+- b_in_octets, ceil(b / 8) for b the output size of H in bits.
+  For example, for b = 256, b_in_octets = 32.
+- k_in_octets, ceil(k / 8) for k the security parameter.
+  For example, for k = 128, k_in_octets = 16.
+
+Input:
+- msg, an octet string.
+- DST, an octet string.
+- len_in_octets, the length of the requested output in octets.
 
 Output:
-- u, an element in F.
+- pseudo_random_octets, an octet string
+
+Notation:
+- For an octet-string str, let str[a : b] represent the
+  (b - a) octets starting at the a'th octet of str.
 
 Steps:
-1. msg_prime = HKDF-Extract(DST, msg || I2OSP(0, 1))
-2. info_pfx = "H2C" || I2OSP(ctr, 1)   # "H2C" is a 3-byte ASCII string
-3. for i in (1, ..., m):
-4.   info = info_pfx || I2OSP(i, 1)
-5.   tv = HKDF-Expand(msg_prime, info, L)
-6.   e_i = OS2IP(tv) mod p
-7. u = (e_1, ..., e_m)
-8. return u
+1. ell = ceil((len_in_octets + k_in_octets) / b_in_octets)
+2. ABORT if ell > 256
```

Comment:
Q: what is the maximum length of input the messages?

Reply:
ell implies a maximum length of the output, namely, `256 * b_in_bytes - k_in_bytes`. For 256-bit hashes this is 8176 bytes.

For comparison, the pairing-friendly curves draft defines BLS48-581. When hashing to G2, we need an element of F_(p^8) for p a 581-bit integer at the 256-bit security level. So L = 105 bytes, and hashing to two elements of F_(p^8) requires `len_in_bytes = 2 * 8 * 105 = 1680` + `k_in_bytes = 32`, so ell = 54 in this case.

(And note that one would use SHA-512, not SHA-256, so ell = 54 is pessimistic.